onadata/apps/api/viewsets/data_viewset.py

```diff
@@ -463,7 +464,7 @@ def get_object(self):
                 raise ParseError(_("Invalid dataid `%(dataid)s`"
                                    % {'dataid': dataid}))
 
-            obj = get_object_or_404(Instance, pk=dataid, xform__pk=pk)
+            return get_object_or_404(Instance, pk=dataid, xform__pk=pk)
```

Comment:
A respectful member of the team would ask for returning `obj` first if the `if` statement is not fulfilled :-P 

Because we spend sometimes to understand what was going on behind the scene, I would add some comments. 
What about something like that 

```python
    def get_object(self) -> Union[XForm, Instance]:
        xform = super().get_object()
        pk_lookup, dataid_lookup = self.lookup_fields
        pk = self.kwargs.get(pk_lookup)
        dataid = self.kwargs.get(dataid_lookup)
        
        if pk is None or dataid is None:
           return  xform 
           
        try:
            int(pk)
        except ValueError:
            raise ParseError(_("Invalid pk `%(pk)s`" % {'pk': pk}))
        try:
            int(dataid)
        except ValueError:
            raise ParseError(_("Invalid dataid `%(dataid)s`"
                               % {'dataid': dataid}))

        return get_object_or_404(Instance, pk=dataid, xform__pk=pk)
```

Reply:
`obj` does return on line 469, but I like your refactoring better anyway 🙂